js/ui/popup.js

```diff
@@ -60,7 +60,14 @@ Popup.prototype = util.inherit(Evented, /** @lends Popup.prototype */{
         this._map = map;
         this._map.on('move', this._update);
         if (this.options.closeOnClick) {
-            this._map.on('click', this._onClickClose);
+            // We attach the `Map#click` listener asynchronously so that it is
+            // possible to open a popup within a `click` event on a child of
+            // the map's container. If we attached the listener syncronously
+            // the popup would close when the click event propogated to `map`.
+            var that = this;
+            setTimeout(function() {
```

Comment:
Is there another way? Ad hoc `setTimeout`s in library code are really dangerous. For example, this will throw if for some reason the popup is removed before the timeout triggers, because `that._map` will be `undefined` after `remove` deletes it.